rllib/models/specs/specs_base.py

```diff
@@ -0,0 +1,209 @@
+
+    @abc.abstractmethod
+    def get_shape(self, tensor: TensorType) -> Tuple[int]:
+        """Returns the shape of a tensor.
+
+        Args:
+            tensor: The tensor whose shape is to be returned.
+
+        Returns:
+            A `tuple` specifying the shape of the tensor.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_dtype(self, tensor: TensorType) -> Any:
+        """Returns the data type of a tensor.
+
+        Args:
+            tensor: The tensor whose data type is to be returned.
+
+        Returns:
+            The data type of the tensor.
+        """
+        raise NotImplementedError
+
+    @DeveloperAPI
+    def sample(self, fill_value: Union[float, int] = 0) -> TensorType:
+        """Samples a tensor with the specified value that is matches the specs.
+
+        Args:
+            fill_value: The value to fill the tensor with.
+
+        Returns:
+            A tensor with the specified value that matches the specs.
+        """
+        sampled_shape = self._sample_shape()
+        return self._sample(sampled_shape, fill_value)
+
+    @abc.abstractmethod
+    def _sample(
+        self, shape: Tuple[int], fill_value: Union[float, int] = 0
+    ) -> TensorType:
+        """Returns a sample tensor with the given shape.
+
+        Args:
+            shape: The shape of the tensor to be sampled.
+            fill_value: The value to fill the tensor with.
+
+        Returns:
+            A tensor with the specified value that matches the specs.
+        """
+        raise NotImplementedError
+
+    def _sample_shape(self) -> Tuple[int]:
+        """Converts the expected shape to a sample shape by replacing TBD dimensions
+        with size of 1."""
+        sampled_shape = tuple()
+        for d in self._expected_shape:
+            if isinstance(d, int):
+                sampled_shape += (d,)
+            else:
+                sampled_shape += (1,)
+        return sampled_shape
+
+    def _parse_expected_shape(self, shape: str, shape_vals: Dict[str, int]) -> tuple:
+        """Converts the input shape to a tuple of integers and strings."""
+        expected_shape = tuple()
+
+        # check the validity of shape_vals and get a list of dimension names
+        d_names = self._validate_shape_vals(shape, shape_vals)
+
+        for d in d_names:
+            d_value = shape_vals.get(d, None)
+            if d_value is None:
+                d_value = d
+            expected_shape += (d if d_value is None else d_value,)
+
+        return expected_shape
+
+    def _validate_shape_vals(self, shape: str, shape_vals: Dict[str, int]) -> List[str]:
+        """Checks if the shape_vals is valid and returns a list of dimension names.
+
+        Valid means that shape consist of unique dimension names and shape_vals only
+        consists of keys that are in shape. Also shape_vals can only contain postive
+        integers.
+        """
+        d_names = shape.split(" ")
```

Comment:
Is `,` a better delimiter than `_`? Not sure but worth thinking about

Reply:
yeah I agree, changing that now, I realized people can make a typo by making two spaces and this is not robust to that. Changing to comma delimiter + ignoring white spaces.